File: q4/q4aSemaphoreImpl.java
import java.util.Random;
import java.util.concurrent.Semaphore;

public class Main {

    public static void main(String args[]) {
        System.out.println("gateway=" + gateway(5));
    }

    public static int gateway(int numReplicas) {
        First first = new First();
        Semaphore semaphore = new Semaphore(0);
        for (int i = 0; i < numReplicas; i++) {
            Request request = new Request(semaphore, first);
            Thread thread = new Thread(request, "request" + i);
            thread.start();
        }

        Timer timer = new Timer(semaphore, first);
        Thread timerThread = new Thread(timer, "timer");
        timerThread.start();
    
        try {
            semaphore.acquire();
        } catch (InterruptedException e) {}
        return first.get();
    }


    public static class Timer implements Runnable {
        private Semaphore semaphore;
        private First first;

        public Timer(Semaphore semaphore, First first) {
            this.semaphore = semaphore;
            this.first = first;
        }

        @Override
        public void run() {
            try {
                Thread.sleep(8 * 1000); // Thread.sleep sleeps milliseconds
            } catch (InterruptedException e) {}
            synchronized(this.first) {
                this.first.set(-1);
                this.semaphore.release();
            }
        }
    }

    public static class Request implements Runnable {
        private Semaphore semaphore;
        private First first;
        private int randomNumber;

        public Request (Semaphore semaphore, First first) {
            this.semaphore = semaphore;
            this.first = first;
            this.randomNumber = (new Random()).nextInt(30) + 1; // Obtain a number between [1 - 30]. 
        }

        public void run() {
            try {
                Thread.sleep(this.randomNumber * 1000); // Thread.sleep sleeps milliseconds
            } catch (InterruptedException e) {}
            this.first.set(this.randomNumber);
            this.semaphore.release();
        }
    }

    public static class First {
        private int value;
        private boolean set;

        public synchronized void set(int value) {
            if (!this.set) {
                this.set = true;
                this.value = value;
            }
        }

        public synchronized int get() {
            return this.value;
        }
    }  

}

File: q4/q4a-semaphore-impl.c
#include <stdio.h>
#include <assert.h>
#include <pthread.h>
#include <semaphore.h>
#include <time.h>

sem_t semaphore;
pthread_mutex_t mutex;
int numberOfThreads = 5;

void *timer (int *first_to_wake_value) {
    printf("Timer will sleep 8 seconds.\n");
    sleep(8); // Sleeps seconds

    pthread_mutex_lock(&mutex);
    if (*first_to_wake_value == 0) {
        *first_to_wake_value = -1;
    }
    sem_post(&semaphore);
    pthread_mutex_unlock(&mutex);
    
    pthread_exit(NULL);
}

void *request (int *first_to_wake_value) {
    int random_number = (rand() % 30) + 1; // // Obtain a number between [1 - 30].
    printf("Request will sleep %d seconds\n", random_number);
    sleep(random_number); // Sleeps seconds

    pthread_mutex_lock(&mutex);
    if (*first_to_wake_value == 0) {
        *first_to_wake_value = random_number;
    }
    sem_post(&semaphore);
    pthread_mutex_unlock(&mutex);

    pthread_exit(random_number);
}

int gateway (int num_replicas) {
    int *first_to_wake_value = 0;
    pthread_t pthreads[num_replicas];
    for (int i = 0; i < num_replicas; i++) {
        pthread_create(&pthreads[i], NULL, &request, &first_to_wake_value);
    }
    pthread_t timerThread;
    //pthread_create(&timerThread, NULL, &timer, &first_to_wake_value);

    sem_wait(&semaphore);
    pthread_mutex_lock(&mutex);
    int valueToReturn = *first_to_wake_value;
    pthread_mutex_unlock(&mutex);
    return valueToReturn;
}

int main (int argc, char *argv[]) {
    srand ( time(NULL) );
    sem_init(&semaphore, 0, 0);
    pthread_mutex_init(&mutex, NULL);
    printf("gateway=%d\n", gateway(numberOfThreads));
}
